index.js

```diff
@@ -70,40 +70,53 @@ function parseDependencies (source, root) {
  * output transformed source.
  */
 function transformSource (runner, config, source, map, callback) {
-  var child = execFile(
+  var child = spawn(
     runner.file,
     runner.arguments.concat(
       runnerPath,
       ioDelimiter,
       config.engine
-    ),
-    { timeout: config.timeoutMs },
-    function (error, stdout, stderr) {
-      // Output is delimited to filter out unwanted warnings or other output
-      // that we don't want in our files.
+    )
+  )
+  if (config.timeoutMs) {
+    var cancelTimeout = setTimeout(function () {
+      child.kill()
+    }, config.timeoutMs)
+  }
+
+  var dataBuffers = []
+  child.stdout.on('data', function (data) {
+    // Output is delimited to filter out unwanted warnings or other output
+    // that we don't want in our files.
+    dataBuffers.push(data)
+  })
+
+  child.on('close', function (code) {
+    if (code === 0) {
       var sourceRegex = new RegExp(ioDelimiter + '([\\s\\S]+)' + ioDelimiter)
-      var matches = stdout.match(sourceRegex)
+      var matches = dataBuffers.join('').match(sourceRegex)
       var transformedSource = matches && matches[1]
-      if (error && error.signal === 'SIGTERM') {
-        if (config.timeoutMs) {
-          callback(new Error(
-            'rails-erb-loader took longer than the specified ' + config.timeoutMs +
-            'ms timeout'
-          ))
-        } else {
-          callback(error)
-        }
-      } else {
-        callback(error, transformedSource, map)
-      }
+      if (config.timeoutMs) { cancelTimeout() }
+      callback(undefined, transformedSource, map)
+    } else if (child.killed) {
+      callback(new Error(
+        'rails-erb-loader took longer than the specified ' + config.timeoutMs +
+        'ms timeout'
+      ))
+    } else {
+      callback(new Error('rails-erb-loader failed with code: ' + code))
     }
-  )
+  })
+
+  child.stderr.on('data', function (data) {
```

Comment:
This will kill the build whenever your Ruby code emits a warning or any logging on stderr, which is not appropriate.

Most likely this should be on the [`error` event](https://nodejs.org/api/child_process.html#child_process_event_error).

```js
child.stderr.on('error', callback)
```

However, note this line from the docs:

> Note: The 'exit' event may or may not fire after an error has occurred. When listening to both the 'exit' and 'error' events, it is important to guard against accidentally invoking handler functions multiple times.

So we might consider preventing this with [`_.once`](https://lodash.com/docs/4.17.5#once):

```js
var once = require('lodash.once')

// ...

function transformSource (runner, config, source, map, callback) {
  var callbackOnce = once(callback)
  // use only `callbackOnce` below...
```

There remains a question of what to do with `stderr` output. I think this should be forwarded to the calling context so that webpack can handle it as normal. This is achievable through the [`stdio` option for `spawn`](https://nodejs.org/api/child_process.html#child_process_options_stdio).

I believe the correct setting is:

```js
 var child = spawn(
     runner.file,
     runner.arguments.concat(
       runnerPath,
       ioDelimiter,
       config.engine
    ),
    { stdio: ['pipe', 'pipe', process.stderr] }
)
```

This should allow read/write of `stdin` and `stdout` via `child.stdin` and `child.stdout` respectively. All error logging however will be logged to console during build.

Reply:
Good call, I hadn't considered warnings in this context. :+1:

Reply:
Looks like we can't listen to `stderr` with as well as set the `stdio` option on spawn, but it looks it's behaving as expected by adding `{ stdio: ['pipe', 'pipe', process.stderr] }` and removing the `child.stderr.on('error', callback)`.

Do we expect to see console/warning output from the rails runner? I don't think that behavior is present in the current implementation.

Reply:
> Looks like we can't listen to stderr with as well as set the stdio option on spawn, but it looks it's behaving as expected by adding { stdio: ['pipe', 'pipe', process.stderr] } and removing the child.stderr.on('error', callback).

That sounds correct to me.

> Do we expect to see console/warning output from the rails runner? I don't think that behavior is present in the current implementation.

Good question! You're right that we're currently ignoring `stderr`, but assuming that information being logged to `stderr` is important it should be surfaced. I think we can pipe it like this and consider adding a flag to suppress it if there are any complaints. My suspicion is that a healthy codebase should not be logging to `stderr` while the runner is going.

Reply:
> Looks like we can't listen to stderr with as well as set the stdio option on spawn, but it looks it's behaving as expected by adding { stdio: ['pipe', 'pipe', process.stderr] } and removing the child.stderr.on('error', callback).

Ah, I see what you're getting at. My apologies, @xicreative, I actually made a mistake [above](https://github.com/usabilityhub/rails-erb-loader/pull/56#discussion_r171739344), the code I meant to show was:

```js
child.on('error', callback)
```

So that an error in the process fails the build. Note the [error event](https://nodejs.org/api/child_process.html#child_process_event_error) will fire called if:

1. The process could not be spawned, or
2. The process could not be killed, or
3. Sending a message to the child process failed.

None of these are currently handled by the `child.on('close'` or `child.stdin.on('error'` _I think_.

I'm generally unsure whether we should be listening to `child.on('close'` or `child.on('exit'`... Do you have a good understanding of consequence of the distinction here? This seems relevant:

> Note that when the 'exit' event is triggered, child process stdio streams might still be open.

> The 'close' event is emitted when the stdio streams of a child process have been closed. This is distinct from the 'exit' event, since multiple processes might share the same stdio streams.

You can see [here](https://github.com/nodejs/node/blob/a8b5192fefe54b140ae923e749104b49c5d3be49/lib/child_process.js#L360) that `execFile` does indeed use the `'close'` event. This appears to mean that the runner process could end, but `stdout` and `stderr` remain open. I'm wondering if this is the underlying cause of #47.

Perhaps the safest course of action is to handle all three child events: `'exit'`, `'close'`  and `'error'` and respond to the first that fires only.

Thoughts?

Reply:
@rhys-vdw 
As I understand it, the `'close'` event relates to the child's io stream, where the `'exit'` event relates to the lifecycle of the child process itself. A `'close'` can occur *after* an `'exit'`, so we probably don't want to act on the `'exit'` since we are concerned with capturing all data from the stdout `'data'` event before triggering the callback.

We could add an additional timeout on the `'exit'` event, in case we're concerned the io stream can get stuck open... I'm not sure if that's a likely, or even a possibility.

Capturing `child.on('error')` sounds like a great idea.

Reply:
Ah, it hadn't occurred to me that `'close'` would happen after `'exit'` unless the stream was forwarded from another process (which I was thinking might have been what spring does). But yes, stick to `'close'` because that's what `execFile` uses. I can play around with isolating the spring problem in a different branch.